prometheus_alerting_consistency.md

```diff
@@ -0,0 +1,105 @@
+
+Clear and actionable alerts are a key component of a smooth operational experience.
+Ensuring you have clear and concise guidelines for engineers and SRE creating new alerts will result in a better experience for end users.
+
+## Recommended Reading
+
+A list of references on good alerting practices:
+
+* [Google SRE Book - Monitoring Distributed Systems][2]
+* [Prometheus Alerting Documentation][3]
+* [Alerting for Distributed Systems][4]
+
+## Style Guide
+
+* Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`
+* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`
+  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
+* Alerts MUST include a `severity` label indicating the alert's urgency.
+  * Valid severities are: `critical` or `warning` — see below for
+    guidelines on writing alerts of each severity.
+* Alerts MUST include `summary` and `description` annotations.
+  * Think of `summary` as the first line of a commit message, or an email
+    subject line.  It should be brief but informative.  The `description` is the
+    longer, more detailed explanation of the alert.
+* Alerts SHOULD include a `namespace` label indicating the source of the alert.
+  * Many alerts will include this by virtue of the fact that their PromQL
+    expressions result in a namespace label.  Others may require a static
+    namespace label.
+* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.
+  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.
+* All alerts MUST include a `sop_url` annotation.
+  * Standard Operating Procedure (SOP) style documentation for resolving alerts is required. (SOP Template Placeholder - https://issues.redhat.com/browse/SIGSRE-55)
+
+## Critical Alerts
+
+Critical alerts are for alerting current and impending disaster situations.  These alerts page an SRE.  The situation should warrant waking someone in the middle of the
+night.
+
+Timeline: ~5 minutes.
+
+Reserve critical level alerts only for reporting conditions that may lead to service unavailability.
+Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.
+Configure critical level alerts so they fire before the situation becomes irrecoverable.
+
+Example critical alert:
+
+```yaml
+- alert: BackendServiceStuck
+  expr: backend_state != 1
+  for: 10m
+  labels:
+    severity: critical
+  annotations:
+    summary: 'The backend service is stuck in a non-ready state'
+    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
```

Comment:
This explains what's going on, but doesn't make it immediately obvious why the SRE is expected to care, or what the SRE is expected to do about it.  Maybe we can assume that "what next?" is covered in the SOP, but I expect the alert description to at least hint at why folks should care, to help them rank several competing alerts when deciding which SOP to read first.

Reply:
@wking What would you suggest as a good description in this example? (Doesn't have to be real, just capture what you're trying to highlight)
If I'm understanding you correctly, my own suggestion would be something like
```
The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes. If this service stays in this state, API calls will fail and users won't be able to manage their resources. Managing resources is a core feature of the service.
```

Reply:
@wking any more thoughts on the above?

Reply:
In the definition of each alert, should we put the SLO that it supports (if there is one) to the description?  Then an SRE would immediately know "why".